src/VisualStudio/Core/Def/RoslynActivityLogger.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    /// <summary>
+    /// Let people to inject <see cref="TraceSource"/> to monitor Roslyn activity
+    /// 
+    /// Here, we don't technically use TraceSource as it is meant to be used. but just as an easy 
+    /// way to log data to listeners.
+    /// 
+    /// this also involves creating string, boxing and etc. so, perf wise, it will impact VS quite a bit.
+    /// this also won't collect trace from Roslyn OOP for now. only in proc activity
+    /// </summary>
+    internal static class RoslynActivityLogger
+    {
+        private static readonly object s_gate = new object();
+
+        public static void SetLogger(TraceSource traceSource)
+        {
+            Contract.ThrowIfNull(traceSource);
+
+            lock (s_gate)
+            {
+                // internally, it just uses our existing ILogger
+                Logger.SetLogger(AggregateLogger.AddOrReplace(new TraceSourceLogger(traceSource), Logger.GetLogger(), l => (l as TraceSourceLogger)?.TraceSource == traceSource));
+            }
+        }
+
+        public static void RemoveLogger(TraceSource traceSource)
+        {
+            Contract.ThrowIfNull(traceSource);
+
+            lock (s_gate)
+            {
+                // internally, it just uses our existing ILogger
+                Logger.SetLogger(AggregateLogger.Remove(Logger.GetLogger(), l => (l as TraceSourceLogger)?.TraceSource == traceSource));
+            }
+        }
+
+        private class TraceSourceLogger : ILogger
+        {
+            private const int LogEventId = 0;
+            private const int StartEventId = 1;
+            private const int EndEventId = 2;
+
+            private static readonly ImmutableDictionary<FunctionId, string> s_functionIdCache;
+
+            public readonly TraceSource TraceSource;
+
+            static TraceSourceLogger()
+            {
+                // build enum to string cache
+                s_functionIdCache =
+                    Enum.GetValues(typeof(FunctionId)).Cast<FunctionId>().ToImmutableDictionary(f => f, f => f.ToString());
+            }
+
+            public TraceSourceLogger(TraceSource traceSource)
+            {
+                TraceSource = traceSource;
+            }
+
+            public bool IsEnabled(FunctionId functionId)
+            {
+                // we log every roslyn activity
+                return true;
```

Comment:
Is this a perf problem in practice?

Reply:
this logger only injected to Roslyn when someone uses project-system-tool to inject it. otherwise, it is no-op. but when it is injected, it does affect perf since it runs in this log everything mode which involves creating bunch of strings. but as how existing project-system-tool works, users are not supposed to enable it all the time, but only enable it while they are collecting data. otherwise, disable which will remove the logger if injected from Roslyn.

Reply:
Because it's just a tool window that people can choose not to show, and it's handy to have installed for when you do run into problems, I expect many people *will* just have the extension installed.  I think we need to be explicit somewhere that logs are being recorded and it will affect performance.

Reply:
installing tool windows doesn't do anything. it is same as existing tool's behavior. user need to click "start" button for logger to actually start recording.

logger will be injected and removed when start button is clicked and when stop button is clicked respectively.